File: 3.longest-substring-without-repeating-characters.py
# @leet start
from collections import deque
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        current_chars = set()
        q = deque()

        global_max = 0
        current_count = 0

        for letter in s:
            if letter in current_chars:
                while letter in current_chars:
                    current_chars.remove(q.popleft())
                    current_count -= 1

            current_chars.add(letter)
            q.append(letter)
            current_count += 1
            global_max = max(current_count, global_max)

        return global_max

# @leet end

File: 3.longest-substring-without-repeating-characters.cpp
// @leet start
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int len = s.size();
        if (len <= 1) {
            return len;
        }

        unordered_set<char> chars;
        int back = 0;
        int result = 1;
        int curr = 1;
        chars.insert(s[0]);

        for (int i = 1; i < len; i++) {
            while(chars.find(s[i]) != chars.end()) {
                chars.erase(s[back]);
                back++;
                curr--;
            }
            chars.insert(s[i]);
            curr++;
            result = max(result, curr);
        }
        return result;
        
    }
};
// @leet end
